firmware/ines.ksy

```diff
@@ -24,73 +36,113 @@ types:
-      - id: zero_fill
-        size: 4
-  f6:
+      - contents: [0, 0, 0, 0, 0]
+    instances:
+      # TODO: Add an enum for mapper. https://wiki.nesdev.com/w/index.php/List_of_mappers
+      mapper:
+        value: f6.lower_mapper | (f7.upper_mapper << 4)
+        doc-ref: https://wiki.nesdev.com/w/index.php/Mapper
+    types:
+      f6:
+        doc-ref: https://wiki.nesdev.com/w/index.php/INES#Flags_6
+        seq:
+          - id: lower_mapper
+            type: b4
+            doc: Lower nibble of mapper number
+          - id: four_screen
+            type: b1
+            doc: Ignore mirroring control or above mirroring bit; instead provide four-screen VRAM
+          - id: trainer
+            type: b1
+            doc: 512-byte trainer at $7000-$71FF (stored before PRG data)
+          - id: has_battery_ram
+            type: b1
+            doc: If on the cartridge contains battery-backed PRG RAM ($6000-7FFF) or other persistent memory
+          - id: mirroring
+            type: b1
+            enum: mirroring
+            doc: if 0, horizontal arrangement. if 1, vertical arrangement
+        enums:
+          mirroring:
+            0: horizontal
+            1: vertical
+      f7:
+        doc-ref: https://wiki.nesdev.com/w/index.php/INES#Flags_7
+        seq:
+          - id: upper_mapper
+            type: b4
+            doc: Upper nibble of mapper number
+          - id: format
+            type: b2
+            doc: If equal to 2, flags 8-15 are in NES 2.0 format
+          - id: playchoice10
+            type: b1
+            doc: Determines if it made for a Nintendo PlayChoice-10 or not
+          - id: vs_unisystem
+            type: b1
+            doc: Determines if it is made for a Nintendo VS Unisystem or not
+      f9:
+        doc-ref: https://wiki.nesdev.com/w/index.php/INES#Flags_9
+        seq:
+          # TODO: enforce zero (similarly to "contents", but on bit level)
+          - type: b7
+          - id: tv_system
+            type: b1
+            enum: tv_system
+            doc: if 0, NTSC. If 1, PAL.
+        enums:
+          tv_system:
+            0: ntsc
+            1: pal
+      f10:
+        doc-ref: https://wiki.nesdev.com/w/index.php/INES#Flags_10
+        seq:
+          # TODO: enforce zero (similarly to "contents", but on bit level)
+          - type: b2
+          - id: bus_conflict
+            type: b1
+            doc: If 0, no bus conflicts. If 1, bus conflicts.
+          # TODO: 0 = true, 1 = false in this case
+          - id: prg_ram
+            type: b1
+            doc: If 0, PRG ram is present. If 1, not present.
+          # TODO: enforce zero (similarly to "contents", but on bit level)
+          - type: b2
+          - id: tv_system
+            type: b2
+            enum: tv_system
+            doc: if 0, NTSC. If 2, PAL. If 1 or 3, dual compatible.
+        enums:
+          tv_system:
+            0: ntsc
+            1: dual
+            2: pal
+            3: dual
```

Comment:
This won't compile for majority of static language targets (like C++, Java, C#), generating something like that:

```java
    public enum TvSystem {
        NTSC(0),
        DUAL(1),
        PAL(2),
        DUAL(3);
// ...
```

which yields something like "error: variable DUAL is already defined in enum TvSystem".

Reply:
Hmm, is there a way to model this better in KSY? Should it be preprocessed before being mapped to an enum?

Reply:
I'd go with `dual1` and `dual2` for the time being. When we'll add writing/generation capabilities, this way you could specify which value you want exactly?